### ISSUE TITLE
Reverse meaning of `Wait()` predicate return value to match `std::condition_variable` and other major C++ libraries

### ISSUE THREAD

**Issue by CodingCanuck** [3rdparty/eventuals#508]
My best understanding of `ConditionVariable::Wait()` is that it takes a predicate that returns `true` if waiting should continue, and returns false when waiting should stop / no more waiting is required. (This is from my reading of the code to handle condition predicates returning true [here](https://github.com/3rdparty/eventuals/blob/cba2c4652e5bce3957034476e09f5b4655ff5e6b/eventuals/lock.h#L615), and from [this unit test](https://github.com/3rdparty/eventuals/blob/cba2c4652e5bce3957034476e09f5b4655ff5e6b/test/lock.cc#L163-L168) which I think specifies a predicate that first returns `true` to wait then later returns `false` to stop waiting).

This meaning of `true`/`false` is reversed compared to other libraries that I know of which offer types like `ConditionVariable`: instead, a predicate usually returns `true` to indicate that *waiting should stop* because some end state has been reached, and `false` means that waiting should continue. IMHO this more common semantics lets you write code that is easier to understand: "wait until a trigger / resume when the following thing happens / block until the following condition is true".

Sample other libraries:
* https://en.cppreference.com/w/cpp/thread/condition_variable/wait
  > `template< class Predicate >void wait( std::unique_lock<std::mutex>& lock, Predicate stop_waiting ); | (2) | (since C++11)`
  > stop_waiting | - | predicate which returns ​false if the waiting should be continued (bool(stop_waiting()) == false).
* https://www.boost.org/doc/libs/1_62_0/libs/fiber/doc/html/fiber/synchronization/conditions.html
  > When the fiber awakens from the wait, then it checks to see if the appropriate condition is now true, and continues if so. If the condition is not true, then the fiber calls wait again to resume waiting.
* https://abseil.io/docs/cpp/guides/synchronization#condvar-condition-variables
  > The call WaitWithTimeout() allows a thread to wait until either a condition is true, or until some time has elapsed. 
* Facebook's folly uses `std::condition_variable` in its `Semaphore`, so it uses the same semantics as above

Let's reverse how we use predicates passed to `Wait()` to use a "wait until a stopping condition" semantic, which would make us more intuitively similar to other C++ synchronization libraries.

Regardless of what semantics we settle on here, it would help a lot to use a more descriptive argument name than `F f` for the predicate. E.g. the [C++ `std::condition_variable`](https://en.cppreference.com/w/cpp/thread/condition_variable/wait) takes a predicate called `stop_waiting`, which pretty clearly means that a return value of `true` will stop waiting. We could adopt a similar name if we switch to more standard `true`/`false` semantics, or choose a name like `keep_waiting` if we keep our current "reversed" semantics.

**Comment by CodingCanuck** [comment]
@benh I'd love your thoughts on this one!

**Comment by benh** [comment]
I always found C++'s "stop condition" confusing, mostly because condition variables are traditionally taught in schools (and are in also done in most other languages) like so:

```cpp
lock.Acquire();
  while (!ready) {
    condition_variable.Wait();
  }
lock.Release();
```

This was the the reason for breaking with `std::condition_variable` semantics; a lot of people in my experience come to C++ condition variables confused when they have to invert the condition.

Rather than require people to know whether it's a "stop" condition or a "while" condition , I like the idea of being explicit and changing the name of `Wait()` to something like `WaitWhile()`.

**Comment by CodingCanuck** [comment]
> I always found C++'s "stop condition" confusing, mostly because condition variables are traditionally taught in schools (and are also done in most other languages) like so:
> 
> ```c++
> lock.Acquire();
> while (!ready) {
>   condition_variable.Wait();
> }
> lock.Release();
> ```

This is a great example: in your snippet, `ready` (or if it was a function / predicate, `ready()`) is exactly what you should pass to standard and common C++ libraries like `std::condition_variable` that expect true-when-done predicates. In contrast, you couldn't pass `ready()` to `::eventuals::Wait()`: you'd instead have to pass in a wrapper or some other predicate like `not_ready()` that returns false-when-done: this flows much less naturally with your example code.

Given that we're writing a C++ library, our APIs will be most intuitive for C++ programmers if we follow C++ standards and norms, rather than those of other languages.

I'm not as familiar with how other languages implement this pattern, so let's do some homework (though other languages should be less relevant to eventuals' design than C++ norms): here's what I've found from a quick google search for this pattern (something like "$LANGUAGE wait predicate" or "$LANGUAGE condition variable predicate" or "$LANGUAGE synchronization primitives") in the 10 most popular languages according to https://www.tiobe.com/tiobe-index/ , skipping inapplicable languages (ASM, SQL) (note that I could've easily found only a subset of relevant APIs here, so these findings are subject to confirmation bias on my part!):

1. Python: [true-when-done style (the C++ norm)](https://docs.python.org/3.7/library/threading.html#threading.Condition.wait_for): 
2. C: Can't find a wait API that takes a predicate (closest is [pthread_cond_wait](https://linux.die.net/man/3/pthread_cond_wait) which is signal based)
3. Java: Can't find a built-in API that takes a predicate (closest is [Condition](https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/locks/Condition.html) which is signal based), but the [popular library recommendation on stackoverflow](https://stackoverflow.com/a/25325830) is [true-when-done style (the C++ norm)](https://github.com/awaitility/awaitility/wiki/Usage#simple)
4. C++: [true-when-done style](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_until)
5. C#: [true-when-done style (the C++ norm)](https://docs.microsoft.com/en-us/dotnet/api/system.threading.spinwait.spinuntil?view=net-6.0#system-threading-spinwait-spinuntil(system-func((system-boolean))))
6. Visual Basic: [true-when-done style (the C++ norm)](https://docs.microsoft.com/en-us/dotnet/api/system.threading.spinwait.spinuntil?view=net-6.0#system-threading-spinwait-spinuntil(system-func((system-boolean))))
7. JavaScript: Can't find a wait API that takes a predicate (closest is [Node.js EventEmitter](https://nodejs.org/api/events.html))
8. Swift: true-when-done style (the C++ norm) ([`XCTWaiter`](https://developer.apple.com/documentation/xctest/xctwaiter/2806744-wait) waits for expectations to be fulfilled, and [`XCTNSPredicateExpectation`](https://developer.apple.com/documentation/xctest/xctnspredicateexpectation) is an expectation that is fulfilled when a predicate returns true)
9. PHP: Can't find a wait API that takes a predicate (closest is [`parallel\Sync::wait`](https://www.php.net/manual/en/parallel-sync.wait.php) that waits until a notification is sent)
10. Go: Can't find a wait API that takes a predicate (closest is [`Cond.Wait`](https://pkg.go.dev/sync#Cond.Wait) that waits until a signal is sent)

Plus some bonus languages not in the top 10:
a. Rust: [false-when-done style (reversed from the C++ norm)](https://doc.rust-lang.org/stable/std/sync/struct.Condvar.html#method.wait_while)
b. Lisp: [true-when-done style (the C++ norm)](http://www.lispworks.com/documentation/lw60/LW/html/lw-1049.htm)

Here's what I found when searching for something like "condition variables $COLLEGE":
1. Stanford: true-when-done style ([one](https://web.stanford.edu/~ouster/cgi-bin/cs140-spring14/lecture.php?topic=locks#:~:text=Condition%20variables%3A%20used%20to%20wait,re%2Dacquire%20lock%20before%20returning.), [two](http://web.stanford.edu/class/archive/cs/cs110/cs110.1204/static/lectures/11-multithreading-cvs-semaphores.pdf) [search for "predicate"] in slides])
2. MIT: can't find predicate examples ([non-predicate example](http://www.mit.edu/people/proven/IAP_2000/cond.html))
3. CMU: can't find good predicate examples ([C++-specific example that references predicates tangentially](https://wiki.sei.cmu.edu/confluence/display/cplusplus/CON54-CPP.+Wrap+functions+that+can+spuriously+wake+up+in+a+loop))
4. Waterloo: true-when-done style ([one](https://student.cs.uwaterloo.ca/~cs350/F16/notes/synch.pdf, search for "condition variable")

Do you have any pointers to languages you were thinking of that use the reversed / false-when-done style, or similarly any courses? It seems to be an unusual pattern, particularly in C++, and Rust was the only example I found of this reversed false-when-done style.

> 
> This was the the reason for breaking with `std::condition_variable` semantics; a lot of people in my experience come to C++ condition variables confused when they have to invert the condition.

Ooh, my experience here is quite opposite: I spent about 1-2 hours yesterday trying to figure out what `Wait()` expected me to pass it, and it took me a long time to convince myself that it was in fact reversed from standard C++ expectations. (Admittedly, comments could've helped a lot here: added in https://github.com/3rdparty/eventuals/pull/507 )

I would similarly expect most C++ programmers who have encountered `Wait()` functions that take primitives to have either encountered standard library functions or those from a super common library like boost or absl, so I'd expect most C++ programmers would find the false-while-done style to be unintuitive.

> 
> Rather than require people to know whether it's a "stop" condition or a "while" condition , I like the idea of being explicit and changing the name of `Wait()` to something like `WaitWhile()`.

Renaming `Wait()` (and its arguments) to make the meaning of its arguments more clear sounds excellent!

If we go with the standard C++ style, `WaitUntil()` or `WaitFor()` would be a more clear name.

The name `WaitUntil` actually reveals that a true-when-done predicate is already a convention in `eventuals`, since that's what you'd provide to [`::eventuals::Until()`](https://github.com/3rdparty/eventuals/blob/cba2c4652e5bce3957034476e09f5b4655ff5e6b/eventuals/until.h#L239) ([sample use in eventuals tests](https://github.com/3rdparty/eventuals/blob/cba2c4652e5bce3957034476e09f5b4655ff5e6b/test/repeat.cc#L31-L33)).

Other than the current implementation of `Wait()`, are there other eventuals that take a false-when-done predicate? (I don't know if true-when-done / C++ normal style or if false-when-done is currently more common in eventuals)

**Comment by benh** [comment]
> Do you have any pointers to languages you were thinking of that use the reversed / false-when-done style, or similarly any courses? It seems to be an unusual pattern, particularly in C++, and Rust was the only example I found of this reversed false-when-done style.

Every language you've mentioned and the course PDFs that you've linked have the semantics that I'd expect, that is, condition variables are traditionally taught and used via an explicit *outer* `while()` loop as in the example I added above:
```cpp
lock.Acquire();
while (!ready) {
  condition_variable.Wait();
}
lock.Release();
```

Many languages do not have condition variables that take predicates, such as C, and Java, and C#, etc, so what people do in those languages looks a lot like the the code snippet above.

As you've pointed out there are many other abstractions that involve waiting and taking a predicate, they just aren't condition variables.

Regardless of whether the abstractions are true condition variables or something else, IMHO the ones that make the semantics clear in the naming of the function are the best ones. You pointed out a Java library for waiting on things via calling `.until()`, and similarly you pointed out a C# abstraction for spinning until some predicate that they call `SpinUntil()`. 

Python's version of _condition variables_ that wait with a predicate is called `wait_for()`. And Rust's version of _condition variables_ that wait with a predicate is called `wait_while()`.

C++ is one of the only languages that I know of that _specifically for condition variables_ let's you wait with a predicate and does not provide some indication in the naming about how that predicate is being used. I repeated that mistake by not properly naming the eventuals version of `Wait()` as I assumed that the more natural inclination for people would be to assume the semantics of the code snippet above versus needing to know to invert the condition! Clearly that was incorrect for at least you! So let's rectify this by making the naming explicit.

> Ooh, my experience here is quite opposite: I spent about 1-2 hours yesterday trying to figure out what Wait() expected me to pass it, and it took me a long time to convince myself that it was in fact reversed from standard C++ expectations.

I'm sorry you had that experience. Were the [tests](https://github.com/3rdparty/eventuals/blob/main/test/lock.cc#L150) helpful at all?
